SVGLength.value doesn't update when writing-mode changes for font-relative units

Ahmad-S792 opened the pull request:
#### 58c53af50a29cac771a859c18c27036fc3696335
<pre>
SVGLength.value doesn&apos;t update when writing-mode changes for font-relative units
<a href="https://bugs.webkit.org/show_bug.cgi?id=303897">https://bugs.webkit.org/show_bug.cgi?id=303897</a>
<a href="https://rdar.apple.com/166190252">rdar://166190252</a>

Reviewed by Vitor Roriz.

When accessing SVGLength.value after changing style attributes like
writing-mode, font-relative units (ch, em, rem, ex) were using stale
computed styles. This caused incorrect values for units like &apos;ch&apos; which
depend on font metrics that vary with writing mode.

The fix ensures style is recalculated before resolving length values,
so font metrics reflect the current computed style.

This is performance neutral on Speedometer and MotionMark. Please refer
to performance testing on radar (Thanks to Taher for help). Additionally,
this also adds `if` conditions to only update stlye when we have length
type, which are relative [e.g., ch] (once again Thanks to Taher for help).
It does so by leveraging lambda which uses switch to list out absolute
units - which at least supported in enum `SVGLengthType` and as per CSS
specification (1-1).

* LayoutTests/imported/w3c/web-platform-tests/svg/types/scripted/SVGLength-ch-expected.txt: Progression
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/types/scripted/SVGLength-ch-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/svg/types/scripted/SVGLength-ch-expected.txt: Removed.
* Source/WebCore/svg/SVGLength.cpp:
(WebCore::SVGLength::valueForBindings):

Canonical link: <a href="https://commits.webkit.org/304657@main">https://commits.webkit.org/304657@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/223b44cded928995c18c0662d7cb83f4ca5d5384

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136200 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8556 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47480 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143909 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ab552040-b60d-4f6c-ab58-748af78a8b24) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138072 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9231 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8401 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104143 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89168 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d3e76f1f-15f0-4a5c-a9b6-00c76d364635) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139146 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6704 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122044 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84971 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9e3cfece-84e3-4c71-a0ca-80599a8630ea) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6374 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4037 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4504 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115657 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40248 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146655 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8240 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40811 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/112492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8256 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6922 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112825 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28637 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6296 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118348 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62227 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8288 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36406 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8006 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71847 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8227 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8080 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

vitorroriz reviewed (commented):
Is there a protected accessor available?

annevk commented on Source/WebCore/svg/SVGLength.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2016-2024 Apple Inc. All rights reserved.
+ * Copyright (C) 2016-2025 Apple Inc. All rights reserved.
```
I believe we shouldn't bump copyright for bug fixes.

brentfulgham replied:
I generally bump them any time we touch a file.

annevk commented on Source/WebCore/svg/SVGLength.cpp:
```diff
@@ -32,7 +32,12 @@ namespace WebCore {
+    // Ensure style is up-to-date for font-relative units (e.g., ch).
+    RefPtr element = contextElement();
+    if (element)
+        element->protectedDocument()->updateStyleIfNeeded();
```
Do other implementations also force layout in this getter or do they achieve the test result in different ways? It doesn't really seem desirable to force more layout to happen in all kinds of style getters.

vitorroriz replied:
I agree with Anne, also, it is probably worth running this change against SP3 given we have a bunch of SVG content there and re-layout here might have some cost.

Ahmad-S792 replied:
@annevk & @vitorroriz - Please check radar, I have updated it with SP3 run from @nullhook and in short, it is neutral.

nullhook replied:
A few notes:

- We shouldn't reflow on absolute units
- I'm not sure which one perfectly fits here: updateStyleIfNeeded vs updateLayoutIgnorePendingStylesheets

vitorroriz commented on Source/WebCore/svg/SVGLength.cpp:
```diff
@@ -32,7 +32,15 @@ namespace WebCore {
+    RefPtr element = contextElement();
+    // Ensure style is up-to-date for font-relative units (e.g., ch).
+    auto lengthType = m_value.lengthType();
+    if (lengthType != SVGLengthType::Number && lengthType != SVGLengthType::Pixels) {
```
I understood we wanted that only for font-relative units, however we are only excluding `::Number` and `::Pixel` here:

```
enum class SVGLengthType : uint8_t {
    Unknown = 0,
    Number,
    Percentage,
    Ems,
    Exs,
    Pixels,
    Centimeters,
    Millimeters,
    Inches,
    Points,
    Picas,
    Lh,
    Ch
};
```

Ahmad-S792 replied:
I think this enum class is aligned with IDL and newer implementation by Taher does not use this at all except for old cases (e.g., it does not have rem but we support it and also relative line height (rlh)). I think for newer cases, it uses CSS implementation, hence, he might have suggested to use only these bits. @nullhook - am I right here in understanding?

Ahmad-S792 replied:
We do have other clean-ups to do here - I mean in SVGLengthValue.

vitorroriz replied:
hmmmmmm, maybe we can make an overload for `CSS::isFontOrRootFontRelativeLength(LengthUnit lengthUnit)` for  `SVGLengthType`, or a conversion from one to the other, or...maybe we don't need this duplicated definition.

brentfulgham replied:
@vitorroriz Do you want that change before we land this?

Ahmad-S792 replied:
The clean-up would be in separate PR otherwise this will become convoluted.

vitorroriz replied:
I agree that the clean up can be on another patch, but I'm not sure we can ignore the other types here. We could create a simple static function isFontOrRootFontRelativeLength that returns true for all font-relative units on SVG.

Ahmad-S792 replied:
Without this `if` check, it is also performance neutral, I can remove `if` case entirely and it wouldn't impact anything. It was just additional.

vitorroriz replied:
Why not a super simple static function `isFontOrRootFontRelativeLength`?

nullhook replied:
> I understood we wanted that only for font-relative units

I know I mentioned reflowing only for font-relative units, but I was wrong. From my findings, we don’t need to reflow for absolute units, which include `::Number` and `::Pixel`. All other types in the SVGLengthType list depend on font metrics, so we should reflow for those.

vitorroriz replied:
@nullhook I'm not familiar with SVGLengthType but aren't these also absolute (just by looking at their names):
`::Points`, `:: Millimeters`, `:: Centimeters`...

vitorroriz replied:
https://www.w3.org/TR/css-values-3/#absolute-lengths

nullhook replied:
fair enough, we shouldn't reflow for the above units then.

Ahmad-S792 replied:
I did mapping between CSS spec and what is in enum and create switch case with lambda in the function to sort out what is needed.

vitorroriz commented on Source/WebCore/svg/SVGLength.cpp:
```diff
@@ -32,7 +32,31 @@ namespace WebCore {
+    RefPtr element = contextElement();
+    // Ensure style is up-to-date for font-relative units (e.g., ch).
+    // https://www.w3.org/TR/css-values-3/#absolute-lengths
+    auto lengthType = m_value.lengthType();
```
nit: we just use this variable once and the name doesn't document anything. I guess we don't need it.

Ahmad-S792 replied:
Should I use it directly?

```
if (requiresStyleUpdate(m_value.lengthType()))
```

vitorroriz replied:
I think that would be better but this is very not important

Ahmad-S792 replied:
I will do it right away, so we can see EWS result with change.

Ahmad-S792 committed:
SVGLength.value doesn't update when writing-mode changes for font-relative units
https://bugs.webkit.org/show_bug.cgi?id=303897
rdar://166190252

Reviewed by Vitor Roriz.

When accessing SVGLength.value after changing style attributes like
writing-mode, font-relative units (ch, em, rem, ex) were using stale
computed styles. This caused incorrect values for units like 'ch' which
depend on font metrics that vary with writing mode.

The fix ensures style is recalculated before resolving length values,
so font metrics reflect the current computed style.

This is performance neutral on Speedometer and MotionMark. Please refer
to performance testing on radar (Thanks to Taher for help). Additionally,
this also adds `if` conditions to only update stlye when we have length
type, which are relative [e.g., ch] (once again Thanks to Taher for help).
It does so by leveraging lambda which uses switch to list out absolute
units - which at least supported in enum `SVGLengthType` and as per CSS
specification (1-1).

* LayoutTests/imported/w3c/web-platform-tests/svg/types/scripted/SVGLength-ch-expected.txt: Progression
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/svg/types/scripted/SVGLength-ch-expected.txt: Removed.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/svg/types/scripted/SVGLength-ch-expected.txt: Removed.
* Source/WebCore/svg/SVGLength.cpp:
(WebCore::SVGLength::valueForBindings):

Canonical link: https://commits.webkit.org/304657@main
